### PR TITLE
Refine

### PR DESCRIPTION
**Pull request by Ledespana**

### CODE REVIEW

File: frontend/components/RefineButton.tsx [hunk 1]
```diff
@@ -1,39 +1,145 @@
...
+        .connect(signer)
+        ["refineToArchive(uint256)"](selectedIterations);
+    } catch (error: any) {
+      toast({
+        title: `Error refining. ${error.message}`,
+        status: "error",
+        position: "bottom-left",
+        duration: 5000,
+        isClosable: true,
+      });
```
Comment (democrazyfi) [inline frontend/components/RefineButton.tsx:59]: cancelling the transaction popup shows a big message in toast
Screenshot 2024-04-04 at 19 34 32

File: frontend/components/RefineButton.tsx [hunk 2]
```diff
@@ -1,39 +1,159 @@
...
-        h="30px"
-        w="70px"
-        fontSize="md"
-        fontFamily={inter.style.fontFamily}
-        fontWeight={700}
-      >
-        REFINE
-      </Button>
+  const refine = async () => {
+    if (isRefining) return;
```
Comment (democrazyfi) [inline frontend/components/RefineButton.tsx:47]:
```suggestion
```

File: frontend/components/RefineButton.tsx [hunk 3]
```diff
@@ -1,39 +1,159 @@
...
+            borderRadius={10}
+            h="30px"
+            w="fit-content"
+            fontSize="md"
+            fontWeight={700}
+            onClick={refine}
+            fontFamily={inter.style.fontFamily}
+          >
+            {isRefining ? "REFINING..." : "REFINE TO ARCHIVE"}
+          </Button>
```
Comment (democrazyfi) [inline frontend/components/RefineButton.tsx:94]:
```suggestion
          <Button
            bg={defaultRed}
            color="white"
            _hover={{ bg: defaultRed }}
            borderRadius={10}
            h="30px"
            w="fit-content"
            fontSize="md"
            fontWeight={700}
            onClick={refine}
            fontFamily={inter.style.fontFamily}
            loadingText={"REFINING..."}
            isLoading={isRefining}
          >
            REFINE TO ARCHIVE
          </Button>
```


If we use loadingText/isLoading we get the benefit of a disabled state

Screenshot 2024-04-05 at 14 01 58

**Review by democrazyfi — APPROVED** [review]
Minor suggestion but lgtm

File: frontend/components/RefineButton.tsx [hunk 4]
```diff
@@ -1,39 +1,159 @@
...
+        ["refineToArchive(uint256)"](selectedIterations);
+      setIsRefining(false);
+    } catch (error: any) {
+      setIsRefining(false);
+      let message = "Error refining: ";
+      if (error.code === "ACTION_REJECTED") {
+        message += "Transaction rejected.";
+      } else if (error.action === "estimateGas") {
+        message +=
+          "Error estimating gas, try with a lower number of iterations.";
```
Comment (tehnalogos) [inline frontend/components/RefineButton.tsx:63]: nice :)

File: frontend/components/RefineButton.tsx (hunk 3 above)
Comment (tehnalogos) [reply frontend/components/RefineButton.tsx:94]: nice thinking @democrazyfi. @Ledespana worth trying the above imo as well.

**Comment by tehnalogos** [comment]
accidentally closed PR, but just reopened. my bad

**Review by tehnalogos — APPROVED** [review]

**Pull request merged**